### DISCUSSION TITLE
iOS Navigation Controller

### DISCUSSION THREAD

**Discussion by Meninblack007** [software-mansion/react-native-screens#2102] [Q&A]
Hi team,

As part of RNScreenStack, I see we are doing both `setViewControllers` and `pushViewController` / `popViewController` while adding or removing a screen from the stack. It had let to crashes for us in the last. Would like to understand why are we doing both?

As per documentation from Apple, `setViewControllers` will handle everything nevertheless.

https://developer.apple.com/documentation/uikit/uinavigationcontroller/1621861-setviewcontrollers#discussion

```
Use this method to update or replace the current view controller stack without pushing or popping each controller explicitly. In addition, this method lets you update the set of controllers without animating the changes, which might be appropriate at launch time when you want to return the navigation controller to a previous state.

If animations are enabled, this method decides which type of transition to perform based on whether the last item in the items array is already in the navigation stack. If the view controller is currently in the stack, but is not the topmost item, this method uses a pop transition; if it is the topmost item, no transition is performed. If the view controller is not on the stack, this method uses a push transition. Only one transition is performed, but when that transition finishes, the entire contents of the stack are replaced with the new view controllers. For example, if controllers A, B, and C are on the stack and you set controllers D, A, and B, this method uses a pop transition and the resulting stack contains the controllers D, A, and B.
```

This is something we have fixed by forking the repo. https://github.com/myntra/react-native-screens/commit/ad2eebe8d8948fcabca80387437a91db4a073644

**Comment by Meninblack007** [comment]
@WoLewicki @kkafar any thoughts on this?

**Reply by kkafar** [reply]
Hi! Answered you [here](https://github.com/software-mansion/react-native-screens/discussions/2102#discussioncomment-9130959)

**Comment by kkafar** [comment]
Yeah, IIRC we use `setViewControllers` in cases where it is possible that react changed more than one controller, and changes could possibly be made deeper in stack (it is possible in react to change whole stack, or something in middle of the stack render-by-render, although I'm not convinced we support it anyway). 
Push / pop is used when there is only a single controller change in play? (writing completely from memory).

Could you elaborate what the issue is exactly? What are the scenarios when the crashes (I suppose) happen?
